modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/ImpliedTrinomialTreeFxSingleBarrierOptionProductPricer.java

```diff
@@ -0,0 +1,470 @@
+    DiscountFactors counterDiscountFactors = ratesProvider.discountFactors(ccyCounter);
+
+    Function<Double, Double> interestRate = new Function<Double, Double>() {
+      @Override
+      public Double apply(Double t) {
+        return counterDiscountFactors.zeroRate(t);
+      }
+    };
+    Function<Double, Double> dividendRate = new Function<Double, Double>() {
+      @Override
+      public Double apply(Double t) {
+        return baseDiscountFactors.zeroRate(t);
+      }
+    };
+    Function<DoublesPair, Double> impliedVolSurface = new Function<DoublesPair, Double>() {
+      @Override
+      public Double apply(DoublesPair tk) {
+        double dfBase = baseDiscountFactors.discountFactor(tk.getFirst());
+        double dfCounter = counterDiscountFactors.discountFactor(tk.getFirst());
+        double forward = todayFx * dfBase / dfCounter;
+        return volatilityProvider.getVolatility(currencyPair, tk.getFirst(), tk.getSecond(), forward);
+      }
+    };
+    ImpliedTrinomialTreeLocalVolatilityCalculator localVol =
+        new ImpliedTrinomialTreeLocalVolatilityCalculator(nSteps, timeToExpiry);
+    return localVol.calibrateImpliedVolatility(impliedVolSurface, todayFx, interestRate, dividendRate);
+  }
+
+  //-------------------------------------------------------------------------
+  private ValueDerivatives priceDerivatives(
+      ResolvedFxSingleBarrierOption option,
+      RatesProvider ratesProvider,
+      BlackVolatilityFxProvider volatilityProvider,
+      RecombiningTrinomialTreeData data) {
+
+    validate(option, ratesProvider, volatilityProvider);
+    validateData(option, ratesProvider, volatilityProvider, data);
+    ResolvedFxVanillaOption underlyingOption = option.getUnderlyingOption();
+    double timeToExpiry = data.getTime(nSteps);
+    ResolvedFxSingle underlyingFx = underlyingOption.getUnderlying();
+    Currency ccyBase = underlyingFx.getCounterCurrencyPayment().getCurrency();
+    Currency ccyCounter = underlyingFx.getCounterCurrencyPayment().getCurrency();
+    DiscountFactors baseDiscountFactors = ratesProvider.discountFactors(ccyBase);
+    DiscountFactors counterDiscountFactors = ratesProvider.discountFactors(ccyCounter);
+    double rebateAtExpiry = 0d; // used to price knock-in option
+    double rebateAtExpiryDerivative = 0d; // used to price knock-in option
+    double notional = Math.abs(underlyingFx.getBaseCurrencyPayment().getAmount());
+    double[] rebateArray = new double[nSteps + 1];
+    SimpleConstantContinuousBarrier barrier = (SimpleConstantContinuousBarrier) option.getBarrier();
+    if (option.getRebate().isPresent()) {
+      CurrencyAmount rebateCurrencyAmount = option.getRebate().get();
+      double rebatePerUnit = rebateCurrencyAmount.getAmount() / notional;
+      boolean isCounter = rebateCurrencyAmount.getCurrency().equals(ccyCounter);
+      double rebate = isCounter ? rebatePerUnit : rebatePerUnit * barrier.getBarrierLevel();
+      if (barrier.getKnockType().isKnockIn()) {
```

Comment:
Explain the in-out parity for barrier including the rebate in the documentation, with the pricing of barrier using trees.